fortanix-vme/fortanix-vme-runner/src/lib.rs

```diff
@@ -286,60 +389,55 @@ impl Server {
                     Direction::Right,
                     "vsock");
                 enclave.write(&serde_cbor::ser::to_vec(&response).unwrap())?;
-                let _ = thread::Builder::new().spawn(move || {
-                    let mut proxy = vsock.connect_with_cid_port(enclave_addr.cid(), enclave_addr.port()).unwrap();
-                    Self::proxy_connection((&mut conn, "remote"), (&mut proxy, "proxy"));
-                });
+
+                let proxy = vsock.connect_with_cid_port(enclave_addr.cid(), enclave_addr.port()).unwrap();
+                server.add_connection(proxy, conn, "remote".to_string());
+
                 Ok(())
             },
             Err(e) => Err(e),
         }
     }
 
-    fn proxy_connection(remote: (&mut TcpStream, &str), proxy: (&mut VsockStream, &str)) {
+    fn proxy_connections(server: Arc<Server>) {
+        let mut closed_connections = Vec::new();
+
         loop {
-            let mut read_set = FdSet::new();
-            read_set.insert(remote.0.as_raw_fd());
-            read_set.insert(proxy.0.as_raw_fd());
-
-            if let Ok(_num) = select(None, Some(&mut read_set), None, None, None) {
-                if read_set.contains(remote.0.as_raw_fd()) {
-                    match Self::transfer_data(remote.0, remote.1, proxy.0, proxy.1) {
-                        Ok(0)  => {
-                            // According to the `Read` threat documentation, reading 0 bytes
-                            // indicates that the connection has been shutdown correctly. So we
-                            // close the proxy service
-                            // https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read
-                            break
-                        },
-                        Ok(_)  => (),
-                        Err(e) => {
-                            eprintln!("transfer from remote failed: {:?}", e);
-                            break;
-                        }
+            // Exchange messages on every proxy connection
+            // TODO: Store connections as a linked hash map so we don't need to keep a read lock
+            // over the HashMap while every connection is serviced
+            if let Ok(connections) = server.connections.read() {
+                for (key, connection) in connections.iter() {
+                    match connection.try_lock() {
+                        Ok(mut connection) => if !connection.proxy() {
+                                            connection.close();
+                                            closed_connections.push(key.clone());
+                                          }
```

Comment:
Nit: wrapping the `if`-block into a separate block might be slightly easier to read:
```
Ok(mut connection) => {
    if !connection.proxy() {
        connection.close();
        closed_connections.push(key.clone());
    }
}
```